apps/ejabberd/src/mongoose_cassandra_worker.erl

```diff
@@ -0,0 +1,494 @@
+        {error, _Reason} ->
+            {noreply, State}
+    end;
+handle_cast({multi_async_cql_query, Module, QueryName, MultiParams},
+    State=#state{conn=Conn}) ->
+    case get_prepared_query(Conn, Module, QueryName, State) of
+        {ok, PreparedQuery, State2} ->
+            [execute_prepared_query(Conn, PreparedQuery, Params)
+             || Params <- MultiParams],
+            {noreply, State2};
+        {error, _Reason} ->
+            {noreply, State}
+    end;
+handle_cast(Msg, State) ->
+    ?WARNING_MSG("Strange cast message ~p.", [Msg]),
+    {noreply, State}.
+
+%%--------------------------------------------------------------------
+%% Function: handle_info(Info, State) -> {noreply, State} |
+%%                                       {noreply, State, Timeout} |
+%%                                       {stop, Reason, State}
+%% Description: Handling all non call/cast messages
+%%--------------------------------------------------------------------
+
+
+handle_info({connection_result, {ok, ConnPid, Conn}},
+            State=#state{conn=undefined}) ->
+    State2 = init_connection(ConnPid, Conn, State),
+    {noreply, State2};
+handle_info({connection_result, Reason}, State=#state{conn=undefined}) ->
+    ?ERROR_MSG("issue=\"Fail to connect to Cassandra\", reason=~1000p", [Reason]),
+    {stop, {connection_result, Reason}, State};
+handle_info(_, State=#state{conn=undefined}) ->
+    {noreply, State};
+handle_info({seestar_response, QueryRef, ResultF}, State) ->
+    {noreply, forward_query_respond(ResultF, QueryRef, State)};
+handle_info({'DOWN', _, process, Pid, Reason}, State=#state{conn=Pid}) ->
+    ?ERROR_MSG("issue=\"Cassandra connection closed\", reason=~1000p", [Reason]),
+    {stop, {dead_connection, Pid, Reason}, State};
+handle_info(Msg, State) ->
+    ?WARNING_MSG("Strange info message ~p.", [Msg]),
+    {noreply, State}.
+
+%%--------------------------------------------------------------------
+%% Function: terminate(Reason, State) -> void()
+%% Description: This function is called by a gen_server when it is about to
+%% terminate. It should be the opposite of Module:init/1 and do any necessary
+%% cleaning up. When it returns, the gen_server terminates with Reason.
+%% The return value is ignored.
+%%--------------------------------------------------------------------
+terminate(_Reason, _State) ->
+    ok.
+
+%%--------------------------------------------------------------------
+%% Func: code_change(OldVsn, State, Extra) -> {ok, NewState}
+%% Description: Convert process state when code is changed
+%%--------------------------------------------------------------------
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%% Helpers
+%%--------------------------------------------------------------------
+
+async_spawn(Addr, Port, ClientOptions) ->
+    ?INFO_MSG("issue=\"Connecting to Cassandra\", address=~p, port=~p",
+              [Addr, Port]),
+    Parent = self(),
+    proc_lib:spawn_link(fun() ->
```

Comment:
Do we really need to spawn a process to start the `seestar_session`? I understand you don't want to start the connection in the `gen_server`'s `init` function.

You could f.e send a message to itself in the `init` function and start the `seestar` connection in relevant handle info. What do you think about that @arcusfelis?